### ISSUE TITLE
Analyze all Jupyter notebooks mentioned in PubMed Central

### ISSUE THREAD

**Issue by Daniel-Mietchen** [sparcopen/open-research-doathon#25]
[Jupyter notebooks](http://jupyter.org/) are a popular vehicle these days to share data science workflows. To get an idea of best practices in this regard, it would be good to analyze a good number of them in terms of their reproducibility and other aspects of usability (e.g. documentation, ease of reuse).

A search in PubMed Central (PMC) reveals the following results:

- [ipynb OR jupyter ](https://www.ncbi.nlm.nih.gov/pmc/?term=ipynb+OR+jupyter)

With currently just 102 hits, a systematic reanalysis seems entirely doable and could perhaps itself be documented by way of reproducible notebooks that might eventually end up being mentioned in PMC.

A good starting point here could be [An open RNA-Seq data analysis pipeline tutorial with an example of reprocessing data from a recent Zika virus study](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4972086/), for which both a [Jupyter notebook](http://nbviewer.jupyter.org/github/maayanlab/Zika-RNAseq-Pipeline/blob/master/Zika.ipynb) and a [Docker image](https://hub.docker.com/r/maayanlab/zika/) are available.

I plan to give a lightning talk on this. Some background is in [this recent news piece](http://www.bbc.com/news/science-environment-39054778).

**Comment by rossmounce** [comment]
A few notes...

With EuropePMC a search for ipynb OR jupyter gives 107 results:
http://europepmc.org/search?query=jupyter%20OR%20ipynb

I find it extremely interesting to note that EuropePMC has the full text for 102 of these 107 articles/preprints

(jupyter OR ipynb) AND (HAS_FT:Y)

Which demonstrates that jupyter/IPython notebooks are almost exclusively associated with open-friendly journals(?) Or perhaps this is a bias influenced by legally-enforced inability to do full text search on 'closed' journals where jupyter/ipynb might be mentioned but can't be found by EuropePMC because they are not allowed.

**Comment by rossmounce** [comment]
Rcode to get bibliographic metadata on each of those 107 hits from EuropePMC:

```
install.packages('europepmc')
library(europepmc)
hits <- epmc_search(query='jupyter%20OR%20ipynb&synonym=TRUE',limit=200)
dim(hits)
names(hits)
write.csv(hits,file="107hits.csv")
```

I've also made available the resulting CSV as an editable spreadsheet via GDocs:
https://docs.google.com/spreadsheets/d/1txg0u9zARHrLkY4MYuz5vmsVCZUOItgybEHqx13Bkbc/edit?usp=sharing 

Perhaps with this sheet we can assign who takes responsibility for which papers?

**Comment by Daniel-Mietchen** [comment]
That's a great starting point &mdash; thanks!

**Comment by npscience** [comment]
+1 from me. Interested to contribute and to see the output.

**Comment by Daniel-Mietchen** [comment]
We've taken Ross' spreadsheet and added some columns for
- Notebook URL
- Code in problem cell
- Problem

The "Code in problem cell" column documents the notebook code causing the first problem, and the "Problem" column gives more details. So far, basically none of the notebooks ran through: We normally stopped after the first such error and went on to the next notebook, but for one rather complex notebook, we tried to go through to the end, which we have not reached yet.

**Comment by Daniel-Mietchen** [comment]
I've also added a column for the PMC URL to reduce the fiddling with URLs.

**Comment by Daniel-Mietchen** [comment]
I notified the Jupyter mailing list: https://groups.google.com/forum/#!topic/jupyter/6pQIarRmrsc .

**Comment by mrw34** [comment]
Here's a write-up of our efforts:

https://markwoodbridge.com/2017/03/05/jupyter-reproducible-science.html

Many thanks to @Daniel-Mietchen for the original idea, and for all the help over the weekend!

**Comment by Daniel-Mietchen** [comment]
@mrw34 Thanks - I'll go right into it.

**Comment by Daniel-Mietchen** [comment]
I found one that actually ran through, albeit after a warning about an old kernel:
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4940747/bin/13742_2016_135_MOESM3_ESM.ipynb . A very simple notebook to test a random number generator, but hey, it works!

To celebrate the event, I introduced color coding to the spreadsheet: red for cases where the run resulted in an error, green when it did not.

**Comment by Daniel-Mietchen** [comment]
Here's a notebook shared *only* as a screenshot, from a paper about reproducibility:
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5014984/figure/figure1/ .

Just added yellow to the spreadsheet for cases like this where the notebook did not produce errors nor ran through, and where "n/a" is not applicable in the sense that there is no notebook.

**Comment by Daniel-Mietchen** [comment]
There is a nice "Ten simple rules" series in PLOS Computational Biology:
http://collections.plos.org/ten-simple-rules . Perhaps we should do one on "how to share Jupyter notebooks"?

They already have [Ten Simple Rules for Reproducible Computational Research](http://dx.doi.org/10.1371/journal.pcbi.1003285) and [Ten Simple Rules for Cultivating Open Science and Collaborative R&D](http://dx.doi.org/10.1371/journal.pcbi.1003244) as well as other somewhat related articles, but none of them seem to touch upon Jupyter notebooks.

**Comment by Daniel-Mietchen** [comment]
Some comments relevant for here are also in https://github.com/sparcopen/open-research-doathon/issues/41#issuecomment-284239044 .

**Comment by Daniel-Mietchen** [comment]
The above close was just as part of the wrap-up of the doathon. I will keep working on it and document my progress over at https://github.com/Daniel-Mietchen/ideas/issues/2 .

**Comment by rossmounce** [comment]
@mrw34 @Daniel-Mietchen excellent write-up!

If licensing allows could you upload somewhere all the .ipynb notebooks you found that were related to those 107 papers?

**Comment by Daniel-Mietchen** [comment]
@rossmounce I plan to do that but haven't yet checked for licensing (added column AH for that).

The notebook URLs are in column AD, which currently has the following list:

* http://nbviewer.jupyter.org/github/biocore/American-Gut/blob/master/ipynb/module2_v1.0.ipynb
* http://nbviewer.jupyter.org/github/Gilles86/in_limbo/blob/master/notebooks/level2_simple.ipynb
* https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4150318/bin/rspb20141025supp2.txt
* http://nbviewer.ipython.org/urls/raw.githubusercontent.com/myGrid/DataHackLeiden/alan/Player_example.ipynb?create=1
* https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3906378/bin/pone.0085777.s003.ipynb
* https://github.com/simkovic/wolfpackRevisited/blob/master/code/Revision1.ipynb
* http://nbviewer.jupyter.org/github/idekerlab/py2cytoscape/blob/develop/examples/New_wrapper_api_sample.ipynb
* http://nbviewer.jupyter.org/github/HullUni-bioinformatics/ReproPhylo/blob/master/Tutorial_files/Git/Git_Demo.ipynb
* http://nbviewer.jupyter.org/github/biocore/American-Gut/blob/master/ipynb/Alpha%20diversity%20notebook.ipynb
* http://nbviewer.jupyter.org/github/biocore/American-Gut/blob/master/ipynb/Power.ipynb
* http://nbviewer.jupyter.org/github/jmumford/orthogonalizaton_ipynb/blob/master/orthogonalization.ipynb
* http://nbviewer.jupyter.org/gist/fabianrost84/3cc58a27b5688f4e2eba
* http://nbviewer.ipython.org/github/germs-lab/frontiers-review-2015/blob/master/frontiers-nb-2015.ipynb
* http://nbviewer.jupyter.org/github/JonM4024/synthetic_lethals/blob/master/double_deletions_example.ipynb
* http://nbviewer.jupyter.org/github/a-slide/iPython-Notebook/blob/master/Notebooks/VCF_analysis.ipynb
* http://nbviewer.jupyter.org/github/shyamrallapalli/fraxinus_version1_data_analysis/blob/master/fraxinus_visits_model/Players-nonplayers.ipynb
* https://duke.app.box.com/s/9kphs2300kumjyowjq06
* https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4669674/bin/12918_2015_236_MOESM5_ESM.ipynb
* http://nbviewer.jupyter.org/github/maayanlab/Zika-RNAseq-Pipeline/blob/master/Zika.ipynb
* https://github.com/ajouary/VR_Zebrafish/blob/master/Code/FishTracking.ipynb
* https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4940747/bin/13742_2016_135_MOESM3_ESM.ipynb
* https://github.com/sulab/myvariant.info/blob/master/docs/ipynb/myvariant_R_miller.ipynb
* https://collab.humanbrainproject.eu/#/collab/704/
* http://goo.gl/UIhBjf
* https://github.com/scanisius/discover-notebooks
* https://github.com/linsalrob/PyFBA/blob/working/iPythonNotebooks/From_functional_roles_to_gap-filling.ipynb; https://github.com/linsalrob/PyFBA/blob/working/iPythonNotebooks/Using_an_SBML_model.ipynb
* https://github.com/mattloose/RUFigs
* https://doi.org/10.5281/zenodo.160333
* http://nbviewer.ipython.org/github/biocore/qiime/blob/1.9.1/examples/ipynb/illumina_overview_tutorial.ipynb
* https://github.com/PeterUlz/FocalAmplifications/tree/master/Focal_amplifications_in_R.ipynb
* https://github.com/LozRiviera/COMPASS_paper
* http://nbviewer.ipython.org/github/biocore/American-Gut/blob/master/ipynb/module2_v1.0.ipynb
* https://github.com/KarchinLab/Tokheim_PNAS_2016
* http://github.com/nickloman/ebov
* http://tinyurl.com/zq6uuja
* http://github.com/linnarsson-lab/ipynb-lamanno2016
* https://github.com/gregcaporaso/office-microbes
* https://github.com/MIT-LCP/mimic-iii-paper/blob/master/notebooks/tsne-vis.ipynb
* https://github.com/isidroc/bioalerts
* https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4980007/bin/pone.0159762.s001.html
* http://github.com/chewgl/uORF_repressiveness_supplemental/
* https://github.com/scipy/scipy/pull/5232
* https://github.com/ipython/docker-notebook
* https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5189966/bin/ppat.1006114.s013.zip
* https://dx.doi.org/10.7554/eLife.12741.029; https://dx.doi.org/10.7554/eLife.12741.030
* https://github.com/biocore/american-gut/blob/master/ipynb/primary-processing/02-filter_sequences_for_blooms.md
* https://github.com/berenslab/pr_bc_connectivity
* https://github.com/Schork-Lab/np_single_nucleus_rnaseq/
* https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4979073/bin/NIHMS798127-supplement-Supplementary_Materials__Appendix_1.html
* Downloadable at http://www.phonetik.phoniatrie.rwth-aachen.de/bkroeger/documents/ipynb_InnerSpeechLoop.zip
* https://github.com/sofroniewn/2pRAM-paper
* Notebook attached as zip file
* Notebook is attached as additional files (in PDF)
* https://github.com/biocore/gneiss/blob/master/ipynb/88soils.ipynb and https://github.com/biocore/gneiss/blob/master/ipynb/cfstudy.ipynb
* https://github.com/chembl/mychembl/blob/master/ipython_notebooks/09_myChEMBL_web_services.ipynb; https://github.com/chembl/mychembl/tree/master/ipython_notebooks
* http://nbviewer.jupyter.org/github/peterwittek/ipython-notebooks/blob/master/Citation_Network_SDP.ipynb.
* http://doi.org/10.5281/zenodo.208230
* http://github.com/ThorntonLab/disease_sims; http://github.com/molpopgen/buRden; http://github.com/molpopgen/fwdpy; http://github.com/molpopgen/TennessenEAonly
* https://github.com/zhuangjun1981/retinotopic_mapping

**Comment by Daniel-Mietchen** [comment]
Mark's write-up is now up at https://markwoodbridge.com/2017/03/05/jupyter-reproducible-science.html .

**Comment by Daniel-Mietchen** [comment]
There is a validator tool for Jupyter notebooks: https://github.com/jupyter/nbformat/blob/master/nbformat/validator.py

**Comment by Daniel-Mietchen** [comment]
[Here](https://groups.google.com/forum/#!topic/jupyter/rKvl5y3IoBo) is a discussion about using Jupyter notebooks programmatically, with a useful [demo](https://gist.github.com/fperez/9716279).

**Comment by Daniel-Mietchen** [comment]
I am thinking of submitting this to [JupyterCon](https://conferences.oreilly.com/jupyter/jup-ny) &mdash; submission deadline March 14. Anyone in?

**Comment by rossmounce** [comment]
@Daniel-Mietchen I'd be happy to help you prepare the abstract submission, do a bit more analysis but I can't go to the meeting :) Does that count as 'in' ?

**Comment by Daniel-Mietchen** [comment]
That's "in enough" for my taste. Don't know whether I can go either, but the point is to reach out to the Jupyter community and to help do something about these issues, e.g. by refining the recommendations and perhaps offering some validation mechanism (think Schematron for XML).

**Comment by mrw34** [comment]
I can help with setting up a minimal "best practice" git repo including a notebook, a requirements file, and a script to automate testing the notebook whenever it's pushed. This would work well with GitLab, who have a free CI and build-status notification service.

By the way, I'm not sure that "validation" has much of a role to play here, at least not in the sense of XML/JSON validation (i.e. ```nbformat```) - we didn't find any notebooks that were syntactically incorrect. I think that execution (with ```nbconvert --execute```) is likely the only way to test the notebooks for the failures we're interested in. Some kind of static analysis would obviously be interesting to categorise problems, but rather hard?

**Comment by Daniel-Mietchen** [comment]
@mrw34 that sounds great. Happy to use GitLab for that. Yes, "validation" is probably the wrong word - just feeling my way into this sphere.

Not sure about categorizing the problems yet, but I am aiming at fixing and documenting all issues with all of these notebooks over the next month or so.

**Comment by rossmounce** [comment]
tbh, should we really expect all 'found' .ipynb notebooks to be reproducible? I think not.

As an author of an .ipynb notebook associated with a preprint:
https://peerj.com/preprints/773/

I can say I _knew_ when I published it that it wouldn't be reproducible by others, but that wasn't the point. The content/data I was working on was under copyright and thus it couldn't be shared with the paper or notebook. Thus I knowingly and purposefully used an .ipynb notebook to document the metadata of what I had access to and the operations I ran on those copyrighted files I had access to, demonstrating transparency and "potential reproducibility" (if the reproducers had legitimate access to the same content I had).

We should probably bear edge-cases like this in mind before judging authors too harshly on the apparent widespread phenomenon of non-reproducible .ipynb notebooks. Full reproducibility is not easy!

**Comment by mrw34** [comment]
Yep, would be interesting to see how many in the list are like that.

Ultimately the key is probably providing a zero-friction way of authors to opt-into continuous verification of notebooks where appropriate.

**Comment by Daniel-Mietchen** [comment]
@rossmounce @mrw34 

I agree - we should not expect all Jupyter notebooks to be reproducible, since there are always exceptions like the ones you outlined. Some of these may be avoidable (e.g. some dependencies on proprietary libraries), others not (e.g. datasets that just are not public).

For me, the point of the exercise was to document the extent to which the notebooks are reproducible (both individually and as a group), as well as the barriers that pop up on the way, and to think about lowering those barriers for notebooks shared in the future.

**Comment by Daniel-Mietchen** [comment]
I am reopening this, since it seems we're still active here.

**Comment by tompollard** [comment]
@Daniel-Mietchen let me know if/how I can assist with this. I'm based in Boston, so could probably make it down to JupyterCon if needed.

**Comment by mrw34** [comment]
Repository created: https://gitlab.com/mwoodbri/jupyter-ci/tree/master

Verification is performed using ```nbconvert --execute```.

**Comment by Daniel-Mietchen** [comment]
@mrw34 Thanks - that looks promising. Have never pushed anything to GitLab, but will try as soon as possible.

How do you deal with "could not find kernel for Python 2" type errors?

**Comment by mrw34** [comment]
@Daniel-Mietchen It assumes Python 3. The author would need to specify Python 2 in the .yaml file if necessary.

A third-party service would need to deduce the correct version from the .Ipynb file, which I think is possible, at least for v4 format notebooks.

**Comment by Daniel-Mietchen** [comment]
Just found out about a project that aims specifically at promoting the use of Jupyter notebooks for reproducible science: 
https://github.com/Reproducible-Science-Curriculum .

**Comment by Daniel-Mietchen** [comment]
While our PubMed search found a number of false positives (i.e. papers with no associated notebook), there are also some false negatives (i.e. papers with an associated notebook that do not show up in the search).

I just found this example: http://nbviewer.jupyter.org/gist/pschloss/9815766/notebook.ipynb , which is associated with https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4139711/ but not mentioned there.

**Comment by mrw34** [comment]
Hmm. Looks like an issue with PubMed version. Source does include reference to notebook: http://www.nature.com/nature/journal/v509/n7500/full/nature13178.html

**Comment by jangondol** [comment]
@mrw34 [says](https://markwoodbridge.com/2017/03/05/jupyter-reproducible-science.html): "The Dockerfile refers to and downloads a couple of tarballs from third-party sources. This is perfectly acceptable, especially as some dependencies may ultimately not be redistributable, and it reduces the size of the source repository, but does assume that these links remain valid in the long term."

Before the web at large fully embraces [content-based addressing](https://github.com/ipfs/ipfs), we could formulate a recommendation to the authors to use the [Wayback Machine](https://archive.org/web/) or a similar mechanism. So when referencing tarballs etc., instead of referring to the live copy on the web (which is subject to bitrot), the authors should refer to a version saved in the Web Archive which is less ephemeral (not completely permanent but better than linking to the live content). There are some downsides (e.g. if people use "naive" upload to web archive via their website, large files will time out) but it could generally work.

But what about those who do NOT follow the recommendation? Well, one approach would be to write a bot (cc: @Daniel-Mietchen) that crawls new notebooks and auto-saves external dependencies to the Web Archive. Another simple run-time tool could be also written that would pre-process Dockerfiles, check for (missing) external dependencies and update the links with the Web Archive reference. Perhaps the [Archive Team](https://github.com/ArchiveTeam) could also help with this?

This is just initial thinking about possible ways to address the issue of broken dependencies. If you like the idea, feel free to open a new issue and develop this further.

**Comment by schmelling** [comment]
I did not read through all of it, but from Mark's blogpost I assume that the issue it not yet closed. 

"There’s no means for notebooks to directly express the dependencies they require."

Sebastian Raschka @rasbt wrote a Jupyter magic extension to document dependencies etc. https://github.com/rasbt/watermark
I use this when I publish my notebooks, see https://github.com/schmelling/reciprocal_BLAST/blob/master/notebooks/2_KaiABC_BLAST_Heatmap.ipynb for example.

I hope that is useful to you.

Cheers,
Nic

**Comment by ketch** [comment]
If you'd like to point to an example of a repo that is doing continuous integration testing of notebooks, here is one (shameless self-promotion):

https://github.com/clawpack/riemann_book

It's not for a paper; we're writing a book in Jupyter notebooks.  We have a reasonably complex chain of dependencies so setting up continuous integration is non-trivial.  We're not currently testing anything for correctness; we're just testing that all notebooks run without errors in Python 2 and 3.  See

https://github.com/clawpack/riemann_book/blob/master/test.py

for the code that executes the notebooks.

**Comment by mrw34** [comment]
@ketch Nice! The scikit-bio cookbook takes a similar approach: https://github.com/biocore/scikit-bio-cookbook

**Comment by rasbt** [comment]
@ketch using something similar as well for my repos (e.g., https://github.com/rasbt/deep-learning-book/blob/master/code/tests/test_notebooks.py) and where I have documentation stored Jupyter Notebooks that gets converted to a Markdown/HTML for the websites. 

I have not experimented with "real" unit tests, yet though. So far, it is just checking whether a cell throws an error or not (which will cause the build to fail via Travis CI). However, it would probably be better to somehow check whether e.g,. Py 2.7 and 3.6 produce the same output in the cells (i.e., wrt integer division and stuff). I guess that's more involved though ... has anyone tried that, yet?

**Comment by ketch** [comment]
@rasbt Sounds like we are at the same stage in this.  There are a couple of packages designed to help test notebooks:

https://github.com/computationalmodelling/nbval
https://github.com/bollwyvl/nosebook
https://gist.github.com/minrk/2620735

I haven't tried any of them yet.

**Comment by rasbt** [comment]
@ketch Thanks for the reference! The script looks need; have to give it a try one day (was almost tempted to write it from scratch ;))

**Comment by Daniel-Mietchen** [comment]
@rossmounce @npscience @mrw34 @tompollard @jangondol @rasbt @ketch @schmelling .
The deadline for submissions to Jupyter Con is about 1 day away, so I will work on it tonight.

Pointers in https://github.com/Daniel-Mietchen/events/blob/master/JupyterCon-2017.md and the associated Gdoc. You're all most welcome to join in.

**Comment by mrw34** [comment]
Another potential forum for presentation: https://www.software.ac.uk/c4rr/cfp

**Comment by Daniel-Mietchen** [comment]
Couldn't get to it earlier, but am in the Gdoc right now to draft a submission (6h left). Happy to look into other contexts for presentation as well.

**Comment by Daniel-Mietchen** [comment]
The PMC / Europe PMC numbers are up to 111 and 113 hits, respectively.

**Comment by Daniel-Mietchen** [comment]
I'm done with a first pass for all form fields. Will read Mark's post again and see whether that gives reason to change anything.

**Comment by Daniel-Mietchen** [comment]
There are currently 73 hits for the Google search
[site:arxiv.org+ipynb](https://www.google.com/search?q=site%3Aarxiv.org+ipynb) .

**Comment by Daniel-Mietchen** [comment]
I have submitted: https://github.com/Daniel-Mietchen/events/blob/master/JupyterCon-2017.md .

**Comment by Daniel-Mietchen** [comment]
An additional article for which Jupyter notebooks are available:
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4428227/ .
It states 

> The empirical and simulated data, as well as the python code used for simulation and analyses, have been deposited in the Dryad repository (datadrayd.org; doi:10.5061/dryad.ht0hs upon publication).

At [that DOI](http://dx.doi.org/10.5061/dryad.ht0hs), there is a ZIP file described as 

> IPython notebooks containing the simulation and analyses code for the manuscript (named simulations.ipynb and analyses.ipynb respectively) and empirical and simulated data used in the manuscript.

**Comment by Daniel-Mietchen** [comment]
The original PMC query

- [ipynb OR jupyter ](https://www.ncbi.nlm.nih.gov/pmc/?term=ipynb+OR+jupyter)

now gives 115 results, which is over 10% growth within less than a month, so our project here seems timely.

One of the additional results is
[Detecting High-Order Epistasis in Nonlinear Genotype-Phenotype Maps](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5340324/), with an accompanying Jupyter notebook in https://github.com/harmslab/notebooks-nonlinear-high-order-epistasis and a Binder version at http://mybinder.org/repo/harmslab/notebooks-nonlinear-high-order-epistasis  , both apparently produced for the preprint version of the paper. I just ran the notebook for their Figure 2, which qualitatively reproduces all three components of [Fig. 2 in the final paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5340324/figure/fig2/), though with interesting deviations in terms of presentation - namely axis scale, axis labels, coloring - that I haven't yet looked at in detail.

I have had good success with re-running Binder versions of notebooks in the past. The first one I really dove into was [this one](https://twitter.com/KyleCranmer/status/698240530900193282) &mdash; couldn't resist to give it another try right now, and it still worked fine.

Some useful background on Binder is in 
http://ivory.idyll.org/blog/2016-mybinder.html
and
http://tritemio.github.io/smbits/2016/03/22/binder-fretbursts/ .

Sadly, a PMC search for

- [mybinder](https://www.ncbi.nlm.nih.gov/pmc/?term=mybinder)

currently does not bring up any results.

**Comment by Daniel-Mietchen** [comment]
A PMC search for

- [docker.com OR "docker image"](https://www.ncbi.nlm.nih.gov/pmc/?term=docker.com+OR+%22docker+image%22)

gives 45 results currently.

**Comment by Daniel-Mietchen** [comment]
I also tried 
http://mybinder.org/repo/margudo/lssgalpy
from 
https://arxiv.org/abs/1702.04268 ,
which gave "failed!", i.e. the container couldn't be created.

**Comment by Daniel-Mietchen** [comment]
Just came across https://twitter.com/o_guest/status/832532349988589568 , which links to https://openlab-flowers.inria.fr/uploads/default/original/1X/65addc14bb2a6a7feaf7690865fa3708d5b0990f.pdf , which contains a set of brief articles around computational reproducibility.

**Comment by khinsen** [comment]
A very nice initiative to put rep*bility to a real test!

Unfortunately I am not surprised by the outcome. For two similar recent experiments (though not about notebooks) with similar results, see https://github.com/ReScience/ReScience/issues/43 and https://f1000research.com/articles/6-124/v1#referee-response-20292.

Note that, contrary to frequent claims, notebooks are not a rep*bility tool at all, but a documentation. In terms of rep*bility, notebooks are exactly equivalent to the Python script that you obtain by stringing together the code cells. They are perhaps even a bit worse, because it is easier (and thus done with less hesitation) to insert shell commands which almost always fail on another machine.

The problems that you encountered are exactly those that motivated my [ActivePapers](http://www.activepapers.org/) project: (1) keeping track of dependencies (code AND data) is too difficult with today's tools, and therefore mistakes are the norm rather than the execption, and (2) for code, knowing the dependencies is not sufficient to get them to work with reasonable effort. For the details see a [recent Webinar](https://github.com/alegrand/RR_webinars/blob/master/7_publications/index.org) about rep*bility issues in scientific publication.

The recommendations given at the end of [your summary](https://markwoodbridge.com/2017/03/05/jupyter-reproducible-science.html) are probably the best one can give today, but they won't fix the problem as long as there are no support tools to help with respecting them. Enabling rep*ble results with reasonable effort from the authors requires a toolchain that encourages/enforces good habits right from the start, long before publication comes into the authors' focus. We don't have such a toolchain today.

An aspect that's worth discussing in the context is that notebooks actually encourage "bad" behavior today. I think that can be fixed, but not before people become aware of it. The fundamental problem with notebooks is that they address two different tasks and in doing so they obscure in the mind of scientists the very different priorities that these two tasks require. One task is performing interactive data analysis, the other task is publishing a polished narrative about this analysis once it is done. You want maximum flexibility in the first task, including "shelling out", using local Python modules under development, and whatever else it takes to explore your data quickly. But you need to get rid of all that when editing a notebook into a narrative, and there is not enough support for this today.

A simple notebook validator could help a lot here: scan a notebook for shell commands and for access to files outside the current working directory, and put a clearly visible warning on the code cells that are affected. One step up would be to scan for Python modules that are not properly installed in some virtual environment. Once you know that all your code is in a single virtual environment and that all your data inside a single directory, it becomes much easier to package everything for publication.

**Comment by ThomasA** [comment]
> Python modules that are not properly installed in some virtual environment.

This was exactly my problem in https://twitter.com/ThomasArildsen/status/844688697920491520

**Comment by khinsen** [comment]
It just occured to me that some of the verification code in the [ActivePapers Python edition](https://github.com/activepapers/activepapers-python) could be adapted to do this kind of scan during the execution of a notebook (or of a plain Python script). In terms of usage, this would require something like

```
import reproducibility
reproducibility.set_project_directory("/path/to/my/data")
```

at the beginning of the script/notebook. Moreover, such a package could provide a list of all Python modules that were used with something like

```
reproducibility.show_used_modules()
```

at the end of the script/notebook. Alternatively, it could raise an exception when modules outside of the current virtual environment are used.

As always with Python, there is no way to make such checks absolutely certain, in particular since they cannot check what happens in extension modules. But it might still be much better than nothing. Opinions?

**Comment by Daniel-Mietchen** [comment]
Thanks, @khinsen and @ThomasA for chiming in here. The pointers in your posts sent me off a long reading chain this morning, which provided me with additional perspectives on these issues &mdash; thanks for that. 

Your ActivePapers project looks interesting, and using something like this for some basic verification would probably be a good step forward &mdash; have you seen @mrw34's [CI for Jupyter verification](https://github.com/sparcopen/open-research-doathon/issues/25#issuecomment-284736196)?

**Comment by ThomasA** [comment]
@khinsen a Ph.D. student of mine, @Chroxvi, has also implemented some functionality in our Magni package that sounds a bit similar: https://github.com/SIP-AAU/Magni/tree/master/magni/reproducibility - http://magni.readthedocs.io/en/latest/magni.reproducibility.html

**Comment by khinsen** [comment]
@Daniel-Mietchen Yes, I have seen this and other CI-based approaches. They are great when applicable but frustrating when they aren't - as all technology-based solutions. For example, @mrw34's approch requires all code dependencies to be pip-installable, all data dependencies to be downloadable or part of the repository, and the whole analysis to run within whatever resource limitations GitLab's CI imposes.

ActivePapers takes an inverse approach to CI: guarantee that if the computation runs to completion, it is also reproducible. This is particularly valuable for long-running computations. Of course ActivePapers has technological limitations as well, in particular the restriction to pure Python code.

The parts that could be useful outside of the full ActivePapers framework are the ones that restrict the modules one is allowed to import and the data files one is allowed to access.

@ThomasA Yes, the code by @Chroxvi explores a similar approach. From the docs it seems to be a bit specialized to your environment. Do you think it could be generalized (not relying on Magni, not relying on Conda)?

**Comment by Daniel-Mietchen** [comment]
@khinsen One of the motivations behind our verification exercise here was to come up with recommendations for how to share software, and Jupyter notebooks in particular.

Having "all data dependencies to be downloadable or part of the repository" seems like a good recommendation to me, and being able to show that your a particular piece of software complies with that criterion is helpful. Yes, there are cases when such an approach is not applicable, but I still think such a recommendation would be a better recommendation than the lack of recommendations that we currently have.

As for pip-installability, this obviously makes sense only for Python dependencies, while Jupyter notebooks can contain code or dependencies in multiple other languages, and there are of course several Python package managers. Still, I think it would be good if more of the code shared through articles in PMC that is pip-installable in principle could actually demonstrate this pip-installability, and if people intending to share their code (e.g. as per https://github.com/BjornFJohansson/pygenome/issues/1 ) were made more aware of these issues and the tools available, the tools and their usage could be improved in response to community needs, e.g. beyond pip or GitLab.

As an aside, I just came across the "Repeatability in Computer Science" project (from 2015) over at http://reproducibility.cs.arizona.edu/ , which set a much lower bar for replicability than we did here but had similar observations: http://reproducibility.cs.arizona.edu/v2/index.html . I assume this is known to some people in this thread - just adding it in here to keep all reproducbility-related information in this repo in one place. I know the thread is getting long and unwieldy, so I'm also thinking of branching it out somewhere, so as to provide a simpler way of getting an overview of the state we are at.

**Comment by khinsen** [comment]
@Daniel-Mietchen First of all, I didn't mean my list of restrictions of @mrw34's approach as a criticism. Every technology has limitations. If there's anything to criticize, it's that the restrictions are not listed explicitly, leaving the task of figuring them out to potential users.

Having all dependencies downloadable or packaged with the notebook is indeed a decent compromise given today's state of the art. Recommending it is OK in a domain where it is most probably applicable. The same goes for pip-installability, although I'd expect its applicability to be limited almost everywhere, given how many Python packages depend on C libraries.

The problem with downloadable data in the long run is that it requires baking in a URL into the notebook. Five years from now, that URL is probably stale. More stable references, such as DOIs, don't permit direct downloading today. So today everyone has to choose between ease of replication and long-term availability. It isn't obvious that one or the other choice is to be preferred in general.

I live in a world where datasets are a few GB in size and processing requires a few hours using 10 to 100 processors on a parallel computer. These machines often have network restrictions that make downloading data from the Internet impossible. I mention this just to illustrate that no recommendations can ever be absolute - there is too much diversity in scientific computing.

**Comment by gbilder** [comment]
Hi. Daniel pointed me at this thread.

When you say that "DOIs, don't permit direct downloading today" I'm not sure if the word "permit" here is in reference to:

- access control restrictions
- the convention that Crossref/DataCite DOIs tend to resolve to a landing page instead of 'the thing itself."

 Or both.

As far as I can tell, parties appear unanimous on the fact that, barring privacy/confidentiality issues, data should be made available openly.  Access control issues should be minimal.

 
And DOIs would allow direct downloading today if, when possible and practical, those registering DOI metadata included links to direct downloads in their Crossref or DataCite metadata. At Crossref we are increasingly seeing publishers registering full text links in their metadata. For examples, see the `text-mining` links in these metadata records from PeerJ:

http://api.crossref.org/members/4443/works

**Comment by khinsen** [comment]
@gbilder My reference was to the landing-page issue. It's good to hear that people are discussion solutions, but as far as I know it remains impractical today to download a dataset automatically given a DOI.

**Comment by mrw34** [comment]
@khinsen @gbilder I guess best-practice here involves assigning DOIs to datasets (and not just their parent publication, if any), and resolving any ambiguity over if/how the data itself can be automatically retrieved given the relevant DOI. Lots more on the latter here: https://jcheminf.springeropen.com/articles/10.1186/s13321-015-0081-7

**Comment by Daniel-Mietchen** [comment]
I was just [pinged](https://twitter.com/npscience/status/847122624870264832) about another validation tool:
- https://github.com/computationalmodelling/nbval

**Comment by Daniel-Mietchen** [comment]
Here's an interesting blog post on what can go wrong in terms of reproducibility (with a focus on R): http://blog.appsilondatascience.com/rstats/2017/03/28/reproducible-research-when-your-results-cant-be-reproduced.html .

**Comment by DCGenomics** [comment]
A demo for the original issue:  https://www.ncbi.nlm.nih.gov/pubmed/27583132
If you click on the LinkOut link, would adding the Jupyter and Docker links here be helpful?

**Comment by mrw34** [comment]
More on reproducibility, including a few Jupyter mentions: https://www.practicereproducibleresearch.org

**Comment by Daniel-Mietchen** [comment]
@DCGenomics I think making the Jupyter, Docker, mybinder etc. versions of the code more discoverable is useful in principle, but conventional LinkOut (which is not shown by default) may not be the best mechanism to do this.

What I could imagine is a mechanism similar to the way images are currently being presented in PubMed, i.e. something that is shown by default if the paper comes with code shared in a standard fashion. That standard would have to be defined, though.

While necessary for reproducibility, discoverability alone is not sufficient, and this example paper highlights that, as explained in Mark's [initial write-up](https://markwoodbridge.com/2017/03/05/jupyter-reproducible-science.html).

**Comment by Daniel-Mietchen** [comment]
I [asked](https://groups.google.com/forum/#!topic/jats4r/SBmnMIQf4xA) for input from the [JATS4R](http://jats4r.org/) community on how such things could/ ought to be tagged in JATS.

**Comment by Daniel-Mietchen** [comment]
There is a [JupyterCon talk about citing Jupyter notebooks](https://conferences.oreilly.com/jupyter/jup-ny/public/schedule/detail/59835). I have contacted the speakers.

**Comment by mpacer** [comment]
I'm not going to read all of this cause it's long. But, this is a cool idea and neat dataset. @eseiver mentioned you wanted to know how to read in notebooks, for which you can use [`nbformat`](https://github.com/jupyter/nbformat) specifically the `nbformat.read()` which you should probably use inside a context manager.

**Comment by Daniel-Mietchen** [comment]
At the [WikiCite 2017](https://meta.wikimedia.org/wiki/WikiCite_2017) hackathon today, we made some further progress in terms of making this analysis itself more reproducible &mdash; a Jupyter notebook that runs the Jupyter notebooks listed in our Google spreadsheet and spits out the first error message: http://paws-public.wmflabs.org/paws-public/995/WikiCite%20notebook%20validator.ipynb .

@mpacer - yes, it makes use of nbformat.read()

**Comment by Daniel-Mietchen** [comment]
At the [WikiCite 2017](https://meta.wikimedia.org/wiki/WikiCite_2017) hackathon today, we made some further progress in terms of making this analysis itself more reproducible &mdash; a Jupyter notebook that runs the Jupyter notebooks listed in our Google spreadsheet and spits out the first error message: http://paws-public.wmflabs.org/paws-public/995/WikiCite%20notebook%20validator.ipynb . @mpacer - yes, it makes use of nbformat.read()

We also looked at Jupyter notebooks cited from Wikipedia &mdash; notes at https://meta.wikimedia.org/wiki/WikiCite_2017/Jupyter_notebooks_on_Wikimedia_sites .

**Comment by yuvipanda** [comment]
Hello! I'm part of the team that's working on beta.mybinder.org and related stuff, and am *extremely* interested in the idea of a 'badge that represents that your code is reproducible, and has been reproduced by CI'. Funnily, I also built the PAWS stuff, which is awesome to find in completely unrelated contexts :D 

The part of the stack I've been attacking right now is the 'how do we reproduce the environment that the analysis took place in', as part of the mybinder work. You can see the project used for that here: https://github.com/jupyter/repo2docker. It takes a git repository and converts it into a Docker image, using conventions that should be easy to use for most people (and does not require them to understand or use Docker unless they want to). It's what powers the building bits of mybinder :)

As part of the CI for that project, you can [see](https://travis-ci.org/jupyter/repo2docker) that we also build and validate some external repositories that are popular! We just represent these as YAML files here: https://github.com/jupyter/repo2docker/tree/master/tests/external and have them auto test on push so we make sure we can keep building them. This can be inverted too - in the repo's CI they can use repo2docker to make sure their changes don't break the build. 

The part where we haven't made much progress yet is in actual validation. nbval mentioned here seems to be the one I like most - it integrates into pytest! We can possibly integrate repo2docker into pytest too, and use that to easily validate repos? Lots of possible avenues to work towards :)

One of the things I'd love to have is something like what https://www.ssllabs.com/ssltest/analyze.html?d=beta.mybinder.org does for HTTPS on websites - scores you on a bunch of factors, with clear ways of improving it. Doing something like that for git repos with notebooks would be great, and I believe we can do a fair amount of work towards it now.

I'll also be at JupyterCon giving a few talks, and would love to meet up if any of you are going to be there!

/ccing @choldgraf who also does a lot of these things with me :)

**Comment by danielskatz** [comment]
Hi!

@Daniel-Mietchen pointed me at this thread/project yesterday, and it seems quite interesting.

I wonder if it makes sense to think about short term and long term reproducibility for notebooks?

By short term, I mean that the notebook might depend on a python package that has to be installed, which could be done by pip before running the notebook, and this step could be automated by a notebook launcher perhaps.

And long term meaning that at some point, the dependency will not work, pip will be replaced by something new, etc., and the only way to solve this is to capture the full environment.  This seems similar to what @yuvipanda describes, and what @tanumalik is trying to do a bit differently in https://arxiv.org/abs/1707.05731 (though I don't think her code is available).  And long term here might still have OS requirements, so maybe I really mean medium term.

Also, I thought I would cc some other people who I think will be interested in this topic, and could perhaps point to other work done in the context of making notebooks reproducible: @fperez @labarba @jennybc @katyhuff

**Comment by ThomasA** [comment]
@khinsen - sorry I lost track of this thread back in March... Yes, I think @Chroxvi's `reproducibility` can be carved out of the Magni package and I actually hope to do that sometime this fall. I hope to do that along with @ppeder08's `validation` (https://github.com/SIP-AAU/Magni/tree/master/magni/utils/validation) which can be used for in-/output-validation of for example somewhat more abstract "data types" than Python's built-in types.

**Comment by TanuMalik** [comment]
Hi Dan,

Our code is available from https://bitbucket.org/geotrust/sciunit-cli

Current documentation is available from
http://geotrusthub.org/geotrust_html/GeoTrust.html

Yes ours is restricted to Linux OS for now. We have a bare bones version
for Mac OS X that is not in production use.

We are currently working on enabling reproducibility for workflows and
Jupyter notebooks through application virtualization. We need some more
work to capture standard I/O.

Tanu

**Comment by Daniel-Mietchen** [comment]
Thanks for the additional comments. I have proposed to work on this further during the Wikimania hackathon: https://phabricator.wikimedia.org/T172848 .

**Comment by Daniel-Mietchen** [comment]
I got sick during the hackathon and haven't fully recoverd, but JupyterCon is just days away, so I have started to distill the discussion here into an outline for the talk next Friday: 
https://github.com/Daniel-Mietchen/events/blob/master/JupyterCon-2017.md#outline-of-the-talk . Will work on it from Tuesday onwards, and your contributions to this are as always welcome.

**Comment by RomanGurinovich** [comment]
After chatting with @Daniel-Mietchen about this idea, we've implemented web app to autorun notebooks mentioned in the paper.

[exe.sci.ai](http://exe.sci.ai/)

Just add list of papers' URLs, like
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5322252/
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4965465/
and run the executability validator.

**It is a pre-pre-pre-alpha version done for fun and in the name of reproducibility. Please, report all the issues and suggest improvements.**

Current setup might require additional horsepower to consume bigger datasets. Also we plan to implement whole repo autodeployment, too many fails because of lack of this feature at the moment.

List of current issues:

https://github.com/sciAI/exe/blob/master/Executability%20issues.md

Validator code:

https://github.com/sciAI/exe

All the credits are going to sci.AI team and, especially @AlexanderPashuk. Alex, thank you for the effort and fights with libs compatibility.

**Comment by RomanGurinovich** [comment]
@yuvipanda, nice job. High five! Dan mentioned, you are on the conference now, right? If you are interested, we can combine effort

**Comment by choldgraf** [comment]
We're both at the conference now and will be at the hack sessions tomorrow!

**Comment by Daniel-Mietchen** [comment]
I was traveling during the hackathon - had heard about it too late. In any case, I hope we can combine efforts with the Binder team. For those interested, the talk sits at 
https://github.com/Daniel-Mietchen/events/blob/master/JupyterCon-2017.md#outline-of-the-talk , and the video shall hopefully become available in a few weeks.

**Comment by Daniel-Mietchen** [comment]
Seen at JupyterCon:
https://github.com/jupyter/repo2docker , a tool that can dockerize a git repo and provide a Jupyter notebook to explore the container.

**Comment by cgpu** [comment]
> Hello! I'm part of the team that's working on beta.mybinder.org and related stuff, and am _extremely_ interested in the idea of a 'badge that represents that your code is reproducible, and has been reproduced by CI'. Funnily, I also built the PAWS stuff, which is awesome to find in completely unrelated contexts :D

Apologies for reviving a closed issue. I am also interested in the reproducibility badge (which is not necessarily the same as the `binder` badge). I came across @mwoodbri 's [jupyter-ci](https://gitlab.com/mwoodbri/jupyter-ci). Any plans for this to be used or people that currently use this? 

(Also cc'ing @yuvipanda as he has been involved in `repo2docker`.)

**Comment by mwoodbri** [comment]
@cgpu Hi! At the time `jupyter-ci` was an answer to "what's the simplest possible way to generate a badge and notify of validation failure?". Simply porting the project to GitHub and GitHub Actions (updating to latest Jupyter best-practice in the process, if necessary) would be a great start. But a more general solution involving `repo2docker` and/or Binder would be even better!

**Comment by cgpu** [comment]
> @cgpu Hi! At the time `jupyter-ci` was an answer to "what's the simplest possible way to generate a badge and notify of validation failure?". Simply porting the project to GitHub and GitHub Actions (updating to latest Jupyter best-practice in the process, if necessary) would be a great start. But a more general solution involving `repo2docker` and/or Binder would be even better!


Hi @mwoodbri , thank you for the prompt response and the background information! I am really fond of the idea of having a binary `does it reproduce` badge of honour, `jupyter-ci` sounds exactly like that.  I am trying to reproduce `.ipynb` accompanied publications as part of a workshop exercise and I am really struggling. It would be nice to know before hands which ones to invest time on, hence the interest. Additionally, it would be nice to have/create a hub or `awesome` repo with the ones that are reproducible only (verified by ci). 

`jupyter-ci` is currently only on gitlab, I know it's the same in essence, but the community is much more active in Github, it would be nice if you have plans to maintain the repo + idea long term to bring it over here. Just a thought.

Thanks once again!

**Comment by mwoodbri** [comment]
@cgpu Here's a version converted to use GitHub Actions: https://github.com/mwoodbri/jupyter-ci

**Comment by cgpu** [comment]
> @cgpu Here's a version converted to use GitHub Actions: https://github.com/mwoodbri/jupyter-ci

@mwoodbri thank you! Time for me to test now :)

@Daniel-Mietchen thank you for providing the reproducibility cafe space for further discussions 2 years after the start of the initiative. Feel free to close this.

**Comment by Daniel-Mietchen** [comment]
Hello everyone. It has been a while since the last post in this thread, but I am happy to report that there is now a preprint that reports on a reproducibility analysis of the Jupyter notebooks associated with publications available via PubMed Central: [Computational reproducibility of Jupyter notebooks from biomedical publications](https://doi.org/10.48550/arXiv.2209.04308) &mdash; joint work with @Sheeba-Samuel . Here is the abstract:

> Jupyter notebooks allow to bundle executable code with its documentation and output in one interactive environment, and they represent a popular mechanism to document and share computational workflows, including for research publications. Here, we analyze the computational reproducibility of 9625 Jupyter notebooks from 1117 GitHub repositories associated with 1419 publications indexed in the biomedical literature repository PubMed Central. 8160 of these were written in Python, including 4169 that had their dependencies declared in standard requirement files and that we attempted to re-run automatically. For 2684 of these, all declared dependencies could be installed successfully, and we re-ran them to assess reproducibility. Of these, 396 notebooks ran through without any errors, including 245 that produced results identical to those reported in the original. Running the other notebooks resulted in exceptions. We zoom in on common problems and practices, highlight trends and discuss potential improvements to Jupyter-related workflows associated with biomedical publications.

For data and code, see  https://doi.org/10.5281/zenodo.6802158 .

I'll keep this thread open until the paper is formally published, and invite your comments in the meantime. Extra pings to some of you who have contributed to this thread before: @mwoodbri @khinsen @yuvipanda @rossmounce @tompollard  @RomanGurinovich @choldgraf @JosephMcArthur .

**Comment by fperez** [comment]
Congrats @Daniel-Mietchen, excellent! I look forward to reading the paper, and will be sure to include it in my reading list for next year's [reproducible research course](https://github.com/UCB-stat-159-s22) I teach at Berkeley!

cc @facusapienza21.

**Comment by leipzig** [comment]
Lots of twitter interest in this preprint. Might be good to dig a bit deeper into all those unknown dependency resolution issues - or perhaps just feature a couple of examples as a panel. There were some comments regarding Docker.

**Comment by khinsen** [comment]
Thanks @Daniel-Mietchen for the update! The preprint is on my e-book reader.

**Comment by yuvipanda** [comment]
Holy shit this is AWESOME!

**Comment by Daniel-Mietchen** [comment]
We're still working on the revision of the paper but here are the slides of our JupyterCon 2023 talk: https://doi.org/10.5281/zenodo.7854503 .

Slide 23  &mdash; How you can get involved &mdash; asks for community input along a number of dimensions, which I am copying here.
- How shall we share the dataset?
- How would you use it?
  - Which variables (article/ repo/ notebook etc.) to look at?
  - Research questions
  - Teaching/ learning opportunities
- What would be useful follow-ups?
- Anyone interested in helping turn this into a service? NumFOCUS?
- Who wants to do this for other languages?
- Would it be useful to publish reports about computational reproducibility?
- Would you like to publish such reports

**Comment by Daniel-Mietchen** [comment]
We recently submitted the revision of the paper &mdash; see https://arxiv.org/abs/2308.07333 for the latest preprint, which describes a complete re-run of our pipeline and provides some more contextualization. In the discussion, we also briefly touch upon scaling issues with such reproducibility studies, mentioning ReScience (pinging @khinsen) as an example. We are keen on putting this dataset to use, e.g. in educational settings (cc @fperez ).

As always, comments etc. are welcome.

**Comment by Daniel-Mietchen** [comment]
Dear all,
thanks for your participation in this thread. 

The paper on this (with @Sheeba-Samuel) was published yesterday: Computational reproducibility of Jupyter notebooks from biomedical publications, https://doi.org/10.1093/gigascience/giad113 .

We remain interested in 
1. ways to integrate our data with research and education workflows around reproducibility
2. collaborating on similar analyses for 
  - other research fields
  - languages other than Python
  - other types of notebooks 
3. exploring how such automated approaches can be integrated with peer review workflows

@mwoodbri @fperez @khinsen @yuvipanda @rossmounce @tompollard @RomanGurinovich @choldgraf @JosephMcArthur .

With that, I am closing this ticket after nearly 7 years - feel free to open up new ones in relevant places to discuss potential follow-ups.

**Comment by rossmounce** [comment]
Extremely impressive feat! Well done @Sheeba-Samuel & @Daniel-Mietchen !
